docs/add-code-flow.md

```diff
@@ -0,0 +1,97 @@
+
+# `UnixfsAPI.Add()`
+*Entrypoint into the `Unixfs` package*
+
+https://github.com/ipfs/go-ipfs/blob/master/core/coreapi/unixfs.go#L78-L86
+
+The `UnixfsAPI.Add()` acts on the input data or files, to build a _merkledag_ node (in essence it is the entire tree represented by the root node) and adds it to the _blockstore_.
+Within the function, a new `Adder` is created with the configured `Blockstore` and __DAG service__`. 
+
+
+# `Adder.AddAllAndPin(files.File)`
+*Entrypoint*
+
+https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/add.go#L427-L431
+
+The more interesting stuff happens in the `Adder.AddAllAndPin(files)` function. 
+
+https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/add.go#L467-L476
+
+Let's focus on the simple case of a single file, handled by `Adder.addFileNode(path, file)` which redirects the case to `Adder.addFile(pathm files.File)`. 
+
+## `Adder.addFile(path, files.File)`
+*Create the _DAG_ and add to `MFS`*
+
+The `addFile(file)` method is responsible taking the data and converting it into a __DAG__ tree, followed by adding the root of the DAG tree in to the `MFS`.
+
+https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/add.go#L508-L521
+
+There are two main methods to focus on -
+
+### `Adder.add(io.Reader)`
+*Create and return the **root** __DAG__ node*
+
+https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/add.go#L115-L137
+
+This method converts the input _data_ (`io.Reader`) to a `DAG` tree. This is done by splitting the data into _chunks_ using the `Chunker` **(HELP: elaborate on chunker types ?)** and organizing them in a `DAG` (with a *trickle* or *balanced* layout). The method returns the **root** of the __DAG__, formatted as an `ipld.Node`.
+
+### `Adder.addNode(ipld.Node, path)`
+*Add **root** __DAG__ node to the `MFS`*
+
+https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/add.go#L365-L399
+
+Now that we have the **root** node of the `DAG`, this needs to be added to the `MFS` file system. 
+The `MFS` **root** is first fetched (or created, if doesn't already exist) by invoking `mfsRoot()`. 
+Assuming the directory structure already exists in the MFS file system, (if it doesn't exist it will be created using `mfs.Mkdir()` function by passing in the `MFS` **root**), the **root** __DAG__ node is added to the `MFS` File system within the `mfs.PutNode()` function.
+
+#### `[MFS] PutNode(mfs.Root, path, ipld.Node)`
+*Insert node at path into given `MFS`*
+
+https://github.com/ipfs/go-mfs/blob/master/ops.go#L101-L113
+
+In this the path is used to determine the `MFS` `Directory`, which is first lookup up in the `MFS` using `lookupDir()` function. This is followed by adding the **root** __DAG__ node (`ipld.Node`) in to the found `Directory` using `directory.AddChild()` method.
+
+#### - `directory.AddChild(filename, ipld.Node)`
```

Comment:
These 3 "add child to directory" subsections could be collapsed into one, we can definitely not mentioned anything HAMT related (using the `BasicDirectory` if we need to give any concrete example). The most important takeaway for the reader should be that MFS handles the UnixFS layer, it adds the child entry in whichever format UnixFS uses and updates its own bookkeeping records.